### PR TITLE
[diffusion]: Fix ZImage SP sharding for caption and latent

### PR DESCRIPTION
**Pull request by dutsc**
## Motivation

The parallel implementation of the zimage model sequence is relatively inefficient because the current implementation pads the frame dimension of the 5D tensor, and the frame dimension of the Zimage model is always 1. Padding leads to increased computation. 
#16407 
#16418 

The optimization strategy for this pull request is to select the maximum of H and W as the effective splitting dimension in the 5-dimensional tensor [B, C, T, H, W], and simultaneously split the caption and input latents across multiple GPUs to achieve sequence-parallel splitting.

## Modifications

- The `shard_latents_for_sp`, `gather_latents_for_sp`, and `post_denoising_loop` methods in the `ZImagePipelineConfig` class have been rewritten to split the larger dimension between H and W.
- Utility functions required by these methods were also added.
- In denoising.py, print the shape of `latent_model_input` to verify the splitting logic.(commented)

## Accuracy Tests

After the modifications, the image generation is now correct and complete compared to before.  

## Benchmarking and Profiling

Generating 1024*1024 size images with 9 denoising steps and H20, the denoising stage P50 latency decreased from **3450ms** on a single card to **1953ms**, resulting in a **1.77x** speedup.

**test environment**: H20, 1024x1024 resolution, 9 denoising steps

| Configuration | P50 Latency (ms) | P90 Latency (ms) | P99 Latency (ms) | Speedup |
| :--- | :---: | :---: | :---: | :---: |
| H20 (Baseline) | 3450 | 3477 | 3527 | 1.0x |
| **2 x H20** | **1953** | **1954** | **2022** |**1.74x-1.77x** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py [hunk 1]
```diff
@@ -70,19 +77,141 @@ def tokenize_prompt(self, prompts: list[str], tokenizer, tok_kwargs) -> dict:
...
+        else:
+            H = int(
+                batch.height // self.vae_config.arch_config.spatial_compression_ratio
+            )
+            W = int(
+                batch.width // self.vae_config.arch_config.spatial_compression_ratio
+            )
+
+        # Rule: shard along the larger spatial dimension (W/H), implemented via optional H/W transpose.
+        # 目的：选择 H 和 W 较大的 进行 shard，所以 H_eff = max(H, W)
```
[the remark this answers is not in the document]
Comment (dutsc) [reply python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py]: It has been translated into English.

File: python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py [hunk 2]
```diff
@@ -70,19 +77,141 @@ def tokenize_prompt(self, prompts: list[str], tokenizer, tok_kwargs) -> dict:
...
+        H_tok_local = H_tok_pad // sp_size
+        h0_tok = rank * H_tok_local
+
+        # Cap/text sharding: avoid duplicating cap tokens across ranks.
+        cap_len = (
+            int(batch.prompt_embeds[0].size(0))
+            if getattr(batch, "prompt_embeds", None)
+            else 0
+        )
+        cap_total = self._ceil_to_multiple(cap_len, 32 * sp_size)
```
[the remark this answers is not in the document]
Comment (dutsc) [reply python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py]: finish

File: python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py [hunk 3]
```diff
@@ -70,19 +77,141 @@ def tokenize_prompt(self, prompts: list[str], tokenizer, tok_kwargs) -> dict:
...
+        H_tok = H_eff // 2
+        pad_tok = plan["H_tok_pad"] - H_tok
+        pad_lat = pad_tok * 2
+        if pad_lat > 0:
+            pad = latents[:, :, :, -1:, :].repeat(1, 1, 1, pad_lat, 1)
+            latents = torch.cat([latents, pad], dim=3)
+
+        h0 = plan["h0_tok"] * 2
+        h1 = (plan["h0_tok"] + plan["H_tok_local"]) * 2
+        latents = latents[:, :, :, h0:h1, :]
```
[the remark this answers is not in the document]
Comment (dutsc) [reply python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py]: done

**Review by yhyang201 — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 4]
```diff
@@ -1017,6 +1017,9 @@ def forward(
 
                         # Predict noise residual
                         attn_metadata = self._build_attn_metadata(i, batch, server_args)
+                        # print(f"latent_model_input shape: {latent_model_input.shape}")
```
Comment (yhyang201) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: remove this print debug code

Comment (dutsc) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Done.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 5]
```diff
@@ -1018,6 +1018,9 @@ def forward(
...
+                        # # [B, C, T, H/sp_size, W] for zimage-turbo
+                        # [B, H*C / sp_size,C] for FLUX.1-dev
```
[the remark this answers is not in the document]
Comment (dutsc) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Done.

**Review by mickqian — COMMENTED** [review]
do we have testcase for this case?
